### PR TITLE
ICS4: Add ORDERED_ALLOW_TIMEOUT channel type

### PR DESCRIPTION
**Pull request by AdityaSripal**
Allow for ORDERED_ALLOW_TIMEOUT, by writing a timeout receipt on destination chain if packet is received after timeout. Once the timeout receipt is written, it can be proven on counterparty chain to timeout packet without closing the channel

### CODE REVIEW

File: spec/core/ics-004-channel-and-packet-semantics/README.md [hunk 1]
```diff
@@ -42,14 +42,17 @@ A *bidirectional* channel is a channel where packets can flow in both directions
 
 A *unidirectional* channel is a channel where packets can only flow in one direction: from `A` to `B` (or from `B` to `A`, the order of naming is arbitrary).
 
-An *ordered* channel is a channel where packets are delivered exactly in the order which they were sent.
+An *ordered* channel is a channel where packets are delivered exactly in the order which they were sent. This channel type offers a very strict guarantee of ordering. Either, the packets are received in the order they were sent, or if a packet in the sequence times out; then all future packets are also not receivable and the channel closes.
```
Comment (ethanfrey) [inline spec/core/ics-004-channel-and-packet-semantics/README.md:45]: This is the current semantics of Ordered Channels? I did not know that.

That definitely makes sense to add a version with looser requirements.

(This will require a hard fork on both sides to add it?)

**Review by ethanfrey — COMMENTED** [review]
Interesting

File: spec/core/ics-004-channel-and-packet-semantics/README.md (hunk 1 above)
Comment (AdityaSripal) [reply spec/core/ics-004-channel-and-packet-semantics/README.md:45]: Correct it will require both sides to have the new type enabled

File: spec/core/ics-004-channel-and-packet-semantics/README.md [hunk 2]
```diff
@@ -929,12 +934,6 @@ function timeoutOnClose(
     // delete our commitment
     provableStore.delete(packetCommitmentPath(packet.sourcePort, packet.sourceChannel, packet.sequence))
 
-    if channel.order === ORDERED {
```
Comment (cwgoes) [inline spec/core/ics-004-channel-and-packet-semantics/README.md:932]: what does this change have to do with the new channel type?

**Review by cwgoes — CHANGES_REQUESTED** [review]
ACK, but see comment

File: spec/core/ics-004-channel-and-packet-semantics/README.md (hunk 2 above)
Comment (AdityaSripal) [reply spec/core/ics-004-channel-and-packet-semantics/README.md:932]: This is just in the wrong place originally.

This is pseudocode for when an ORDERED channel times out. But that is already specified on line 856.

But this function is for when a channel has already closed, and we want to time out all pending packets. So there is no  special case for ORDERED channels here.

File: spec/core/ics-004-channel-and-packet-semantics/README.md [hunk 3]
```diff
@@ -624,7 +628,7 @@ function recvPacket(
 
     // all assertions passed (except sequence check), we can alter state
 
-    if (channel.order === ORDERED) {
+    if (channel.order === ORDERED || channel.order == ORDERED_ALLOW_TIMEOUT) {
       nextSequenceRecv = provableStore.get(nextSequenceRecvPath(packet.destPort, packet.destChannel))
```
Comment (colin-axner) [inline spec/core/ics-004-channel-and-packet-semantics/README.md:632]: I don't understand. If packet sequence 5 times out, even if the channel doesn't close, I don't see any changes here that allow packet sequence 6 to be received

Comment (AdityaSripal) [reply spec/core/ics-004-channel-and-packet-semantics/README.md:632]: Yes great catch, this makes the changes less trivial as we need to somehow communicate to destination chain the sequences that have been timed out

File: spec/core/ics-004-channel-and-packet-semantics/README.md [hunk 4]
```diff
@@ -859,6 +859,11 @@ function timeoutPacket(
       channel.state = CLOSED
       provableStore.set(channelPath(packet.sourcePort, packet.sourceChannel), channel)
     }
+    // on ORDERED_ALLOW_TIMEOUT, increment NextSequenceRecv so that next packet sequence can be received once this packet times out.
+    if channel.order === ORDERED_ALLOW_TIMEOUT {
+      nextSequenceRecv = nextSequenceRecv + 1
+      provableStore.set(nextSequenceRecvPath(packet.destPort, packet.destChannel), nextSequenceRecv)
```
Comment (colin-axner) [inline spec/core/ics-004-channel-and-packet-semantics/README.md:865]: this should be the nextSequenceAck (as it is on the sender side)

**Comment by AdityaSripal** [comment]
As colin mentioned this feature will be more complicated than first envisioned.

The ack sequence can be correctly incremented when a timeout occurs. However, the receive sequence needs to be incremented on the counterparty chain.

Thus, the receiving chain needs to have this information propogated back to it somehow. This will probably require some slightly different relayer behaviour to enable.

e.g. We might have to relay the ack sequence to the receiving chain before being able to receive the next packet after timeout. Though, async acks complicates this

File: spec/core/ics-004-channel-and-packet-semantics/README.md [hunk 5]
```diff
@@ -610,8 +624,31 @@ function recvPacket(
     abortTransactionUnless(connection !== null)
     abortTransactionUnless(connection.state === OPEN)
 
-    abortTransactionUnless(packet.timeoutHeight === 0 || getConsensusHeight() < packet.timeoutHeight)
-    abortTransactionUnless(packet.timeoutTimestamp === 0 || currentTimestamp() < packet.timeoutTimestamp)
+    switch channel.order {
+      case ORDERED:
```
Comment (colin-axner) [inline spec/core/ics-004-channel-and-packet-semantics/README.md:628]: Is it possible to do something like? I get that this is a typescript fall through, but coming from a go lens it performs completely different behaviour 
```go
case ORDERED || UNORDERED:
```
or
```go
case ORDERED, UNORDERED:
```

File: spec/core/ics-004-channel-and-packet-semantics/README.md [hunk 6]
```diff
@@ -610,8 +624,31 @@ function recvPacket(
     abortTransactionUnless(connection !== null)
     abortTransactionUnless(connection.state === OPEN)
 
-    abortTransactionUnless(packet.timeoutHeight === 0 || getConsensusHeight() < packet.timeoutHeight)
-    abortTransactionUnless(packet.timeoutTimestamp === 0 || currentTimestamp() < packet.timeoutTimestamp)
+    switch channel.order {
```
Comment (colin-axner) [inline spec/core/ics-004-channel-and-packet-semantics/README.md:627]: I imagine the logic like so:
```
if packet.TimedOut() {
    switch channel.Order {
    case ORDERED_ALLOW_TIMEOUT:
        // write timeout reciept
    default:
        abort()
    }   
}
```
Main request is that logically, we are only performing certain logic if the packet has timed out

File: spec/core/ics-004-channel-and-packet-semantics/README.md [hunk 7]
```diff
@@ -823,37 +869,64 @@ function timeoutPacket(
...
+        abortTransactionUnless(nextSequenceRecv <= packet.sequence)
+        // ordered channel: check that the recv sequence is as claimed
+        abortTransactionUnless(connection.verifyNextSequenceRecv(
+          proofHeight,
+          proof,
+          packet.destPort,
+          packet.destChannel,
+          nextSequenceRecv
+        ))
+        break;
```
Comment (colin-axner) [inline spec/core/ics-004-channel-and-packet-semantics/README.md:884]: what's the break for?

File: spec/core/ics-004-channel-and-packet-semantics/README.md [hunk 8]
```diff
@@ -823,37 +869,64 @@ function timeoutPacket(
...
+    // only close on strictly ORDERED channels
     if channel.order === ORDERED {
       // ordered channel: close the channel
       channel.state = CLOSED
       provableStore.set(channelPath(packet.sourcePort, packet.sourceChannel), channel)
     }
+    // on ORDERED_ALLOW_TIMEOUT, increment NextSequenceAck so that next packet can be acknowledged after this packet timed out.
+    if channel.order === ORDERED_ALLOW_TIMEOUT {
+      nextSequenceAck = nextSequenceAck + 1
+      provableStore.set(nextSequenceAckPath(packet.destPort, packet.destChannel), nextSequenceAck)
```
Comment (colin-axner) [inline spec/core/ics-004-channel-and-packet-semantics/README.md:928]: This should be sourcePort/sourceChannel

**Review by colin-axner — COMMENTED** [review]
Concept ACK. Wonderful work. Clever solution

File: spec/core/ics-004-channel-and-packet-semantics/README.md [hunk 9]
```diff
@@ -624,22 +661,31 @@ function recvPacket(
...
-      )
-    }
+    switch channel.order {
+      case ORDERED:
+      case ORDERED_ALLOW_TIMEOUT:
+        nextSequenceRecv = provableStore.get(nextSequenceRecvPath(packet.destPort, packet.destChannel))
+        abortTransactionUnless(packet.sequence === nextSequenceRecv)
+        nextSequenceRecv = nextSequenceRecv + 1
+        provableStore.set(nextSequenceRecvPath(packet.destPort, packet.destChannel), nextSequenceRecv)
+        break;
```
Comment (mpoke) [inline spec/core/ics-004-channel-and-packet-semantics/README.md:671]:
```suggestion
        // for ORDERED_ALLOW_TIMEOUT, write the sentinel timeout value in packet receipt
        if (channel.order == ORDERED_ALLOW_TIMEOUT  && getConsensusHeight() > packet.timeoutHeight && packet.timeoutHeight != 0) || (currentTimestamp() > packet.timeoutTimestamp && packet.timeoutTimestamp != 0) {
          packetReceipt = TIMEOUT_RECEIPT
        } 
        break
```

File: spec/core/ics-004-channel-and-packet-semantics/README.md [hunk 10]
```diff
@@ -624,22 +661,31 @@ function recvPacket(
 
     // all assertions passed (except sequence check), we can alter state
```
Comment (mpoke) [inline spec/core/ics-004-channel-and-packet-semantics/README.md:663]:
```suggestion
    // assume the packet is successfully received 
    packetReceipt = SUCCESSFUL_RECEIPT
```

File: spec/core/ics-004-channel-and-packet-semantics/README.md (hunk 5 above)
Comment (mpoke) [reply spec/core/ics-004-channel-and-packet-semantics/README.md:628]: I'd use the first variant, i.e., `case ORDERED || UNORDERED`. Seems less ambiguous.

File: spec/core/ics-004-channel-and-packet-semantics/README.md [hunk 11]
```diff
@@ -624,22 +661,31 @@ function recvPacket(
...
+      case UNORDERED:
+        // for unordered channels we must set the receipt so it can be verified on the other side
+        // this receipt does not contain any data, since the packet has not yet been processed
+        // it's the sentinel success receipt: []byte{0x01}
+        abortTransactionUnless(provableStore.get(packetReceiptPath(packet.destPort, packet.destChannel, packet.sequence) === null))
+        provableStore.set(
+          packetReceiptPath(packet.destPort, packet.destChannel, packet.sequence),
+          SUCCESFUL_RECEIPT
+        )
+      break;
```
Comment (mpoke) [inline spec/core/ics-004-channel-and-packet-semantics/README.md:682]:
```suggestion
      break
```
Set the PacketReceipt only once after the switch.

File: spec/core/ics-004-channel-and-packet-semantics/README.md [hunk 12]
```diff
@@ -624,22 +661,31 @@ function recvPacket(
...
 
+      default:
+        // unsupported channel type
+        abortTransactionUnless(true)
+    }
+
```
Comment (mpoke) [inline spec/core/ics-004-channel-and-packet-semantics/README.md:688]:
```suggestion
    provableStore.set(
      packetReceiptPath(packet.destPort, packet.destChannel, packet.sequence),
      packetReceipt
    )
    if (packetReceipt == TIMEOUT_RECEIPT) {
      return
    }
```

File: spec/core/ics-004-channel-and-packet-semantics/README.md [hunk 13]
```diff
@@ -624,22 +661,31 @@ function recvPacket(
...
+          SUCCESFUL_RECEIPT
+        )
+      break;
 
+      default:
+        // unsupported channel type
+        abortTransactionUnless(true)
+    }
+    
     // log that a packet has been received
```
Comment (mpoke) [inline spec/core/ics-004-channel-and-packet-semantics/README.md:689]: As a side note about ICS4, shouldn't we call here the `OnRecvPacket()` callback function and in case it returns a non-nil ACK, invoke `writeAcknowledgement`?

File: spec/core/ics-004-channel-and-packet-semantics/README.md [hunk 14]
```diff
@@ -823,37 +869,64 @@ function timeoutPacket(
...
-        proofHeight,
-        proof,
-        packet.destPort,
-        packet.destChannel,
-        packet.sequence
-      ))
+    switch channel.order {
+      case ORDERED:
+        // ordered channel: check that packet has not been received
+        abortTransactionUnless(nextSequenceRecv <= packet.sequence)
```
Comment (mpoke) [inline spec/core/ics-004-channel-and-packet-semantics/README.md:875]: Is not related to the `ORDERED_ALLOW_TIMEOUT` channels, but shouldn't this condition be `abortTransactionUnless(nextSequenceRecv == packet.sequence)`? Why do we allow timeout messages regarding future packets. Maybe the packet with  `nextSequenceRecv` will still be received on time, even if another packet with e.g.,  `nextSequenceRecv+1` has timed out.

File: spec/core/ics-004-channel-and-packet-semantics/README.md [hunk 15]
```diff
@@ -823,37 +869,64 @@ function timeoutPacket(
...
+          packet.destPort,
+          packet.destChannel,
+          packet.sequence
+        ))
+        break;
+
+      // NOTE: For ORDERED_ALLOW_TIMEOUT, the relayer must first attempt the receive on the destination chain
+      // before the timeout receipt can be written and subsequently proven on the sender chain in timeoutPacket
+      case ORDERED_ALLOW_TIMEOUT:
+        abortTransactionUnless(connection.verifyPacketReceipt(
```
Comment (mpoke) [inline spec/core/ics-004-channel-and-packet-semantics/README.md:900]: In order for the following property to hold, we need to check for the sequence here, similar to the ORDERED case.
> On *ordered_allow_timeout* channels, packets should be sent and received in the same order: if packet *x* is sent before packet *y* by a channel end on chain `A`, packet *x* must be received **or** timed out before packet *y* by the corresponding channel end on chain `B`.

This means that a packet with sequence e.g., 5 cannot be timed out while we are still waiting for an ACK from sequence  e.g., 3

**Review by mpoke — CHANGES_REQUESTED** [review]
Cool idea. Great work. See my comments below.

File: spec/core/ics-004-channel-and-packet-semantics/README.md [hunk 16]
```diff
@@ -610,8 +624,31 @@ function recvPacket(
...
+        // then return
+        if (getConsensusHeight() > packet.timeoutHeight && packet.timeoutHeight != 0) || (currentTimestamp() > packet.timeoutTimestamp && packet.timeoutTimestamp != 0) {
+          nextSequenceRecv = nextSequenceRecv + 1
+          provableStore.set(nextSequenceRecvPath(packet.destPort, packet.destChannel), nextSequenceRecv)
+          provableStore.set(
+          packetReceiptPath(packet.destPort, packet.destChannel, packet.sequence),
+            TIMEOUT_RECEIPT
+          )
+        }
+        return;
```
[the remark this answers is not in the document]
Comment (AdityaSripal) [reply spec/core/ics-004-channel-and-packet-semantics/README.md:654]: reordered

File: spec/core/ics-004-channel-and-packet-semantics/README.md (hunk 10 above)
Comment (AdityaSripal) [reply spec/core/ics-004-channel-and-packet-semantics/README.md:662]: The packet is successfully received at this point. There's no assumption being made

File: spec/core/ics-004-channel-and-packet-semantics/README.md (hunk 14 above)
Comment (AdityaSripal) [reply spec/core/ics-004-channel-and-packet-semantics/README.md]: Hmm yea that makes sense to me.

File: spec/core/ics-004-channel-and-packet-semantics/README.md (hunk 10 above)
Comment (mpoke) [reply spec/core/ics-004-channel-and-packet-semantics/README.md:662]: What I meant by the comment is that we set a variable to `SUCCESSFUL_RECEIPT` to avoid multiple if conditions below, i.e.,`packetReceipt =  SUCCESSFUL_RECEIPT` unless stated otherwise.

File: spec/core/ics-004-channel-and-packet-semantics/README.md [hunk 17]
```diff
@@ -823,37 +869,68 @@ function timeoutPacket(
...
-        proofHeight,
-        proof,
-        packet.destPort,
-        packet.destChannel,
-        packet.sequence
-      ))
+    switch channel.order {
+      case ORDERED:
+        // ordered channel: check that packet has not been received
+        // only allow timeout on next sequence so all sequences before the timed out packet can be received
```
Comment (mpoke) [inline spec/core/ics-004-channel-and-packet-semantics/README.md:875]: This comment is a bit unclear. Did you mean to say "all sequences **AFTER** the timed out packet can be received"?

File: spec/core/ics-004-channel-and-packet-semantics/README.md [hunk 18]
```diff
@@ -823,37 +869,68 @@ function timeoutPacket(
...
+          packet.destChannel,
+          packet.sequence
+        ))
+        break;
+
+      // NOTE: For ORDERED_ALLOW_TIMEOUT, the relayer must first attempt the receive on the destination chain
+      // before the timeout receipt can be written and subsequently proven on the sender chain in timeoutPacket
+      case ORDERED_ALLOW_TIMEOUT:
+        // ordered channel: check that packet has not been received
+        // only allow timeout on next sequence so all sequences before the timed out packet can be received
```
Comment (mpoke) [inline spec/core/ics-004-channel-and-packet-semantics/README.md:902]: ditto

File: spec/core/ics-004-channel-and-packet-semantics/README.md [hunk 19]
```diff
@@ -823,37 +869,68 @@ function timeoutPacket(
...
+        abortTransactionUnless(nextSequenceRecv == packet.sequence)
+        abortTransactionUnless(connection.verifyPacketReceipt(
```
Comment (mpoke) [inline spec/core/ics-004-channel-and-packet-semantics/README.md:904]: Shouldn't `verifyPacketReceipt` also get the sequence as param?

File: spec/core/ics-004-channel-and-packet-semantics/README.md [hunk 20]
```diff
@@ -823,37 +869,68 @@ function timeoutPacket(
...
+        abortTransactionUnless(connection.verifyPacketReceiptAbsence(
+          proofHeight,
+          proof,
+          packet.destPort,
+          packet.destChannel,
+          packet.sequence
+        ))
+        break;
+
+      // NOTE: For ORDERED_ALLOW_TIMEOUT, the relayer must first attempt the receive on the destination chain
```
Comment (mpoke) [inline spec/core/ics-004-channel-and-packet-semantics/README.md:898]: Not sure I understand why is this necessary. Why can't the relayer just send a timeout packet to the sender chain and handle the check as in the ORDERED case, i.e., using the `verifyNextSequenceRecv` method?

**Review by mpoke — APPROVED** [review]
Nice work. See my comments below; mostly nits. One thing worth looking into is that I don't find the reasoning for the `TIMEOUT_RECEIPT` packet receipt sentinel value to be clear enough.
